openwisp_controller/geo/approximate_location/test_approximate_location.py

```diff
@@ -0,0 +1,298 @@
+                org_settings_obj.full_clean()
+            try:
+                self.assertEqual(
+                    context_manager.exception.message_dict[
+                        "approximate_location_enabled"
+                    ][0],
+                    "Approximate Location feature requires "
+                    + "WHOIS Lookup feature to be enabled.",
+                )
+            except AssertionError:
+                self.fail("ValidationError message not equal to expected message.")
+
+        with self.subTest(
+            "Test Approximate Location field visible on admin when "
+            "WHOIS_CONFIGURED True"
+        ):
+            self._login()
+            org = self._get_org()
+            url = reverse(
+                "admin:openwisp_users_organization_change",
+                args=[org.pk],
+            )
+            response = self.client.get(url)
+            self.assertContains(
+                response, 'name="config_settings-0-approximate_location_enabled"'
+            )
+
+        with override_settings(
+            OPENWISP_CONTROLLER_WHOIS_GEOIP_ACCOUNT=None,
+            OPENWISP_CONTROLLER_WHOIS_GEOIP_KEY=None,
+        ):
+            importlib.reload(config_app_settings)
+            with self.subTest(
+                "Test Approximate Location field hidden on admin when "
+                "WHOIS_CONFIGURED False"
+            ):
+                self._login()
+                org = self._get_org()
+                url = reverse(
+                    "admin:openwisp_users_organization_change",
+                    args=[org.pk],
+                )
+                response = self.client.get(url)
+                self.assertNotContains(
+                    response, 'name="config_settings-0-approximate_location_enabled"'
+                )
+
+
+class TestApproximateLocationTransaction(
+    TestApproximateLocationMixin, WHOISTransactionMixin, TransactionTestCase
+):
+    _WHOIS_GEOIP_CLIENT = (
+        "openwisp_controller.config.whois.tasks.geoip2_webservice.Client"
+    )
+
+    def setUp(self):
+        super().setUp()
+        self.admin = self._get_admin()
+
+    @mock.patch.object(config_app_settings, "WHOIS_CONFIGURED", True)
+    @mock.patch(
+        "openwisp_controller.geo.approximate_location.tasks.manage_approximate_locations.delay"  # noqa
+    )
+    @mock.patch(_WHOIS_GEOIP_CLIENT)
+    def test_approximate_location_task_called(
+        self, mocked_client, mocked_approximate_location_task
+    ):
+        connect_whois_handlers()
+        mocked_client.return_value.city.return_value = self._mocked_client_response()
+        self._task_called(
+            mocked_approximate_location_task, task_name="Approximate location"
+        )
+
+        Device.objects.all().delete()
+        device = self._create_device()
+        with self.subTest(
+            "Approximate location task called when last_ip has related WhoIsInfo"
+        ):
```

Comment:
Not sure on this but if there is only one sub test, should we create sub tests?

Reply:
There are other subtests called as well in statement above it https://github.com/openwisp/openwisp-controller/blob/fde52b0fc53cdfc6dcb967b2cb696cfbe4f32256/openwisp_controller/geo/approximate_location/test_approximate_location.py#L121